# Imports

In [1]:
import pandas as pd
from datetime import datetime

# Load data

In [2]:
df = pd.read_csv('../data/ILEDA_anonymized.csv')
df

Institution                                  Course actor.id  \
0              UEF  Advanced Data Management Systems (THJ)    aeb0c   
1              UEF  Advanced Data Management Systems (THJ)    aeb0c   
2              UEF  Advanced Data Management Systems (THJ)    aeb0c   
3              UEF  Advanced Data Management Systems (THJ)    aeb0c   
4              UEF  Advanced Data Management Systems (THJ)    aeb0c   
...            ...                                     ...      ...   
306740         BMU                                   IT131    2db36   
306741         BMU                                   IT131    fc5fe   
306742         BMU                                   IT131    d0f73   
306743         BMU                                   IT131    dfea0   
306744         BMU                                   IT131    7f889   

                   timestamp                                  verb.id  \
0       2023-07-07T18:55:47Z      http://id.tincanapi.com/verb/viewed   
1       2023-07-07T15:21:18Z      http://id.tincanapi.com/verb/viewed   
2       2023-07-07T15:08:13Z      http://id.tincanapi.com/verb/viewed   
3       2023-07-07T15:08:10Z      http://id.tincanapi.com/verb/viewed   
4       2023-07-07T09:03:32Z      http://id.tincanapi.com/verb/viewed   
...                      ...                                      ...   
306740  2023-06-16T16:52:55Z    http://adlnet.gov/expapi/verbs/scored   
306741  2023-06-16T17:58:19Z    http://adlnet.gov/expapi/verbs/scored   
306742  2023-06-16T17:08:43Z    http://adlnet.gov/expapi/verbs/scored   
306743  2023-06-16T17:43:28Z    http://adlnet.gov/expapi/verbs/scored   
306744  2023-06-09T16:23:38Z  http://adlnet.gov/expapi/verbs/answered   

              verb.display                             object.definition.name  \
0                   viewed             Advanced Data Management Systems (THJ)   
1                   viewed                  Assignment Instructions: ER & EER   
2                   viewed                           Recording: Intro lecture   
3                   viewed             Advanced Data Management Systems (THJ)   
4                   viewed             Advanced Data Management Systems (THJ)   
...                    ...                                                ...   
306740  attained grade for  Lekcija 7 - IP adresiranje i sabnetovanje - Tr...   
306741  attained grade for  Lekcija 8 - Transportni sloj - Tradicionalna n...   
306742  attained grade for  Lekcija 9 - Principi rada TCP protokola - Trad...   
306743  attained grade for   IT131 - L04 (638138597640325722 - Tradicionalno)   
306744            answered                               IT131-T02-Pitanje 15   

                                   object.definition.type  \
0         http://id.tincanapi.com/activitytype/lms/course   
1                http://adlnet.gov/expapi/activities/link   
2                http://adlnet.gov/expapi/activities/link   
3         http://id.tincanapi.com/activitytype/lms/course   
4         http://id.tincanapi.com/activitytype/lms/course   
...                                                   ...   
306740     http://adlnet.gov/expapi/activities/assessment   
306741     http://adlnet.gov/expapi/activities/assessment   
306742     http://adlnet.gov/expapi/activities/assessment   
306743     http://adlnet.gov/expapi/activities/assessment   
306744  http://adlnet.gov/expapi/activities/cmi.intera...   

                                                  context  \
0                                              UEF Moodle   
1       UEF Moodle > Advanced Data Management Systems ...   
2       UEF Moodle > Advanced Data Management Systems ...   
3                                              UEF Moodle   
4                                              UEF Moodle   
...                                                   ...   
306740                                      LAMS > 220870   
306741                                      LAMS > 221218   
306742                

# Duplicate values

In [3]:
(df.shape)

(306745, 12)

In [4]:
df.shape[0] - df.drop_duplicates().shape[0]

7454

There are 7454 duplicate rows in the dataset, which we have to drop.

In [5]:
df.drop_duplicates(inplace=True)
df.shape

(299291, 12)

# Dataset statistics

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 299291 entries, 0 to 306739
Data columns (total 12 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Institution             299291 non-null  object 
 1   Course                  299291 non-null  object 
 2   actor.id                299291 non-null  object 
 3   timestamp               299291 non-null  object 
 4   verb.id                 299291 non-null  object 
 5   verb.display            299291 non-null  object 
 6   object.definition.name  299291 non-null  object 
 7   object.definition.type  299291 non-null  object 
 8   context                 299291 non-null  object 
 9   result.score.scaled     11558 non-null   float64
 10  result.success          299291 non-null  bool   
 11  result.completion       299291 non-null  bool   
dtypes: bool(2), float64(1), object(9)
memory usage: 25.7+ MB


In [7]:
df.describe(include='all')

Institution                                   Course actor.id  \
count       299291                                   299291   299291   
unique           4                                        8      829   
top            UEF  Ihminen ja vuorovaikutteinen teknologia    76d01   
freq        153290                                   134855     6255   
mean           NaN                                      NaN      NaN   
std            NaN                                      NaN      NaN   
min            NaN                                      NaN      NaN   
25%            NaN                                      NaN      NaN   
50%            NaN                                      NaN      NaN   
75%            NaN                                      NaN      NaN   
max            NaN                                      NaN      NaN   

                   timestamp                              verb.id  \
count                 299291                               299291   
unique                245548                                   10   
top     2023-06-04T23:09:00Z  http://id.tincanapi.com/verb/viewed   
freq                      67                               229721   
mean                     NaN                                  NaN   
std                      NaN                                  NaN   
min                      NaN                                  NaN   
25%                      NaN                                  NaN   
50%                      NaN                                  NaN   
75%                      NaN                                  NaN   
max                      NaN                                  NaN   

       verb.display                   object.definition.name  \
count        299291                                   299291   
unique           10                                     1549   
top          viewed  Ihminen ja vuorovaikutteinen teknologia   
freq         229721                                    34133   
mean            NaN                                      NaN   
std             NaN                                      NaN   
min             NaN                                      NaN   
25%             NaN                                      NaN   
50%             NaN                                      NaN   
75%             NaN                                      NaN   
max             NaN                                      NaN   

                                 object.definition.type  \
count                                            299291   
unique                                               15   
top     http://id.tincanapi.com/activitytype/lms/course   
freq                                             102682   
mean                                                NaN   
std                                                 NaN   
min                                                 NaN   
25%                                                 NaN   
50%                                                 NaN   
75%                                                 NaN   
max                                                 NaN   

                                                  context  \
count                                              299291   
unique                                                189   
top     UEF Moodle > Ihminen ja vuorovaikutteinen tekn...   
freq                                                45880   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

        result.score.scaled result.success result.completion  
count           11558.00000         299

In [8]:
df.groupby(df.columns.tolist(), as_index=False).size()

Institution                        Course actor.id  \
0             BMU                         CS120    05aba   
1             BMU                         CS120    05aba   
2             BMU                         CS120    05aba   
3             BMU                         CS120    05aba   
4             BMU                         CS120    05aba   
...           ...                           ...      ...   
11553          UL  ARQUITECTURA DE COMPUTADORES    fa4d0   
11554          UL  ARQUITECTURA DE COMPUTADORES    fa4d0   
11555          UL  ARQUITECTURA DE COMPUTADORES    fa4d0   
11556          UL  ARQUITECTURA DE COMPUTADORES    fee26   
11557          UL  ARQUITECTURA DE COMPUTADORES    fee26   

                  timestamp                                   verb.id  \
0      2023-03-18T18:55:20Z   http://adlnet.gov/expapi/verbs/answered   
1      2023-03-18T18:55:20Z   http://adlnet.gov/expapi/verbs/answered   
2      2023-03-18T18:55:20Z   http://adlnet.gov/expapi/verbs/answered   
3      2023-03-18T18:55:20Z   http://adlnet.gov/expapi/verbs/answered   
4      2023-03-18T18:55:20Z   http://adlnet.gov/expapi/verbs/answered   
...                     ...                                       ...   
11553  2023-06-01T16:59:31Z  http://adlnet.gov/expapi/verbs/completed   
11554  2023-06-02T10:47:07Z     http://adlnet.gov/expapi/verbs/scored   
11555  2023-06-07T09:15:26Z  http://adlnet.gov/expapi/verbs/completed   
11556  2023-02-23T17:11:26Z  http://adlnet.gov/expapi/verbs/completed   
11557  2023-06-02T10:44:09Z     http://adlnet.gov/expapi/verbs/scored   

             verb.display                      object.definition.name  \
0                answered                        CS120-T01-Pitanje 20   
1                answered                        CS120-T01-Pitanje 25   
2                answered                        CS120-T01-Pitanje 28   
3                answered                        CS120-T01-Pitanje 39   
4                answered                         CS120-T01-Pitanje 4   
...                   ...                                         ...   
11553           completed                            Lab#6 Defense G4   
11554  attained grade for                           Lab Assignment #6   
11555           completed             Defence quizz lab Assignment #4   
11556           completed  Initial evaluation quizz lab assignment #1   
11557  attained grade for                           Lab Assignment #6   

                                  object.definition.type  \
0      http://adlnet.gov/expapi/activities/cmi.intera...   
1      http://adlnet.gov/expapi/activities/cmi.intera...   
2      http://adlnet.gov/expapi/activities/cmi.intera...   
3      http://adlnet.gov/expapi/activities/cmi.intera...   
4      http://adlnet.gov/expapi/activities/cmi.intera...   
...                                                  ...   
11553             http://xapi.jisc.ac.uk/activities/quiz   
11554     http://adlnet.gov/expapi/activities/assessment   
11555             http://xapi.jisc.ac.uk/activities/quiz   
11556             http://xapi.jisc.ac.uk/activities/quiz   
11557     http://adlnet.gov/expapi/activities/assessment   

                                                 context  result.score.scaled  \
0                     LAMS > test > Pitanje 20 > Attempt             0.000000   
1                     LAMS > test > Pitanje 25 > Attempt             1.000000   
2                     LAMS > test > Pitanje 28 > Attempt             1.000000   
3                     LAMS > test > Pitanje 39 > Attempt             1.000000   
4                      LAMS > test > Pitanje 4 > Attempt             1.000000   
...                                                  ...                  ...   
11553  New Site > ARQUITECTURA DE COMPUTADORES - Grad...             0.560000   
11554  New Site > ARQUITECTURA DE COMPUTADORES - Grad...             0.560000   
11555  New Site > ARQUITECTURA DE COMPUTADORES - Grad...             0.708333   


# Columns

## Verb

In [9]:
df['verb.id'].value_counts()

verb.id
http://id.tincanapi.com/verb/viewed           229721
http://adlnet.gov/expapi/verbs/answered        33546
http://adlnet.gov/expapi/verbs/completed       18224
http://activitystrea.ms/schema/1.0/receive      6582
http://activitystrea.ms/schema/1.0/start        4400
http://adlnet.gov/expapi/verbs/scored           4236
http://activitystrea.ms/schema/1.0/submit       1481
http://activitystrea.ms/schema/1.0/create        931
http://activitystrea.ms/schema/1.0/join          136
http://activitystrea.ms/schema/1.0/leave          34
Name: count, dtype: int64

In [10]:
df['verb.id'] = df['verb.id'].str.split('/').str[-1]

In [11]:
df

Institution                                  Course actor.id  \
0              UEF  Advanced Data Management Systems (THJ)    aeb0c   
1              UEF  Advanced Data Management Systems (THJ)    aeb0c   
2              UEF  Advanced Data Management Systems (THJ)    aeb0c   
3              UEF  Advanced Data Management Systems (THJ)    aeb0c   
4              UEF  Advanced Data Management Systems (THJ)    aeb0c   
...            ...                                     ...      ...   
306723         BMU                                   IT131    87441   
306724         BMU                                   IT131    1c33c   
306730         BMU                                   IT131    98607   
306734         BMU                                   IT131    98607   
306739         BMU                                   IT131    00c5e   

                   timestamp    verb.id verb.display  \
0       2023-07-07T18:55:47Z     viewed       viewed   
1       2023-07-07T15:21:18Z     viewed       viewed   
2       2023-07-07T15:08:13Z     viewed       viewed   
3       2023-07-07T15:08:10Z     viewed       viewed   
4       2023-07-07T09:03:32Z     viewed       viewed   
...                      ...        ...          ...   
306723  2023-06-11T22:57:27Z   answered     answered   
306724  2023-05-16T00:05:45Z   answered     answered   
306730  2023-05-12T22:10:50Z  completed    completed   
306734  2023-02-27T22:45:33Z  completed    completed   
306739  2023-06-01T21:23:16Z   answered     answered   

                                   object.definition.name  \
0                  Advanced Data Management Systems (THJ)   
1                       Assignment Instructions: ER & EER   
2                                Recording: Intro lecture   
3                  Advanced Data Management Systems (THJ)   
4                  Advanced Data Management Systems (THJ)   
...                                                   ...   
306723                               IT131-T05-Pitanje 13   
306724                               IT131-T03-Pitanje 15   
306730                                          IT131-T03   
306734  Lekcija 2 - Osnove umrežavanja - Tradicionalna...   
306739                               IT131-T04-Pitanje 20   

                                   object.definition.type  \
0         http://id.tincanapi.com/activitytype/lms/course   
1                http://adlnet.gov/expapi/activities/link   
2                http://adlnet.gov/expapi/activities/link   
3         http://id.tincanapi.com/activitytype/lms/course   
4         http://id.tincanapi.com/activitytype/lms/course   
...                                                   ...   
306723  http://adlnet.gov/expapi/activities/cmi.intera...   
306724  http://adlnet.gov/expapi/activities/cmi.intera...   
306730     http://adlnet.gov/expapi/activities/assessment   
306734    http://id.tincanapi.com/activitytype/lms/module   
306739  http://adlnet.gov/expapi/activities/cmi.intera...   

                                                  context  \
0                                              UEF Moodle   
1       UEF Moodle > Advanced Data Management Systems ...   
2       UEF Moodle > Advanced Data Management Systems ...   
3                                              UEF Moodle   
4                                              UEF Moodle   
...                                                   ...   
306723                 LAMS > test > Pitanje 13 > Attempt   
306724                 LAMS > test > Pitanje 15 > Attempt   
306730                                        LAMS > test   
306734                                      LAMS > 217371   
306739                 LAMS > test > Pitanje 20 > Attempt   

        result.score.scaled  result.success  result.completion  
0                       NaN           False              False  
1                       NaN           False              False  
2                       NaN           False              False  
3                     

In [12]:
all(df['verb.id'].str.split('/').str[-1] == df['verb.display'])

False

In [13]:
df[df['verb.id'].str.split('/').str[-1] != df['verb.display']][['verb.id', 'verb.display']]

verb.id        verb.display
74       scored  attained grade for
96       scored  attained grade for
97       scored  attained grade for
98       scored  attained grade for
99       scored  attained grade for
...         ...                 ...
306087    start             started
306278  receive            received
306292  receive            received
306398    start             started
306439    start             started

[17800 rows x 2 columns]

In [14]:
df[['verb.id', 'verb.display']].drop_duplicates()

verb.id        verb.display
0          viewed              viewed
29      completed           completed
74         scored  attained grade for
714        submit           submitted
742        create             created
763       receive            received
892      answered            answered
906         start             started
153743       join              joined
153749      leave                left

Each ID corresponds to a single display.

In [15]:
df.drop(['verb.display'], axis=1, inplace=True)

In [16]:
df

Institution                                  Course actor.id  \
0              UEF  Advanced Data Management Systems (THJ)    aeb0c   
1              UEF  Advanced Data Management Systems (THJ)    aeb0c   
2              UEF  Advanced Data Management Systems (THJ)    aeb0c   
3              UEF  Advanced Data Management Systems (THJ)    aeb0c   
4              UEF  Advanced Data Management Systems (THJ)    aeb0c   
...            ...                                     ...      ...   
306723         BMU                                   IT131    87441   
306724         BMU                                   IT131    1c33c   
306730         BMU                                   IT131    98607   
306734         BMU                                   IT131    98607   
306739         BMU                                   IT131    00c5e   

                   timestamp    verb.id  \
0       2023-07-07T18:55:47Z     viewed   
1       2023-07-07T15:21:18Z     viewed   
2       2023-07-07T15:08:13Z     viewed   
3       2023-07-07T15:08:10Z     viewed   
4       2023-07-07T09:03:32Z     viewed   
...                      ...        ...   
306723  2023-06-11T22:57:27Z   answered   
306724  2023-05-16T00:05:45Z   answered   
306730  2023-05-12T22:10:50Z  completed   
306734  2023-02-27T22:45:33Z  completed   
306739  2023-06-01T21:23:16Z   answered   

                                   object.definition.name  \
0                  Advanced Data Management Systems (THJ)   
1                       Assignment Instructions: ER & EER   
2                                Recording: Intro lecture   
3                  Advanced Data Management Systems (THJ)   
4                  Advanced Data Management Systems (THJ)   
...                                                   ...   
306723                               IT131-T05-Pitanje 13   
306724                               IT131-T03-Pitanje 15   
306730                                          IT131-T03   
306734  Lekcija 2 - Osnove umrežavanja - Tradicionalna...   
306739                               IT131-T04-Pitanje 20   

                                   object.definition.type  \
0         http://id.tincanapi.com/activitytype/lms/course   
1                http://adlnet.gov/expapi/activities/link   
2                http://adlnet.gov/expapi/activities/link   
3         http://id.tincanapi.com/activitytype/lms/course   
4         http://id.tincanapi.com/activitytype/lms/course   
...                                                   ...   
306723  http://adlnet.gov/expapi/activities/cmi.intera...   
306724  http://adlnet.gov/expapi/activities/cmi.intera...   
306730     http://adlnet.gov/expapi/activities/assessment   
306734    http://id.tincanapi.com/activitytype/lms/module   
306739  http://adlnet.gov/expapi/activities/cmi.intera...   

                                                  context  \
0                                              UEF Moodle   
1       UEF Moodle > Advanced Data Management Systems ...   
2       UEF Moodle > Advanced Data Management Systems ...   
3                                              UEF Moodle   
4                                              UEF Moodle   
...                                                   ...   
306723                 LAMS > test > Pitanje 13 > Attempt   
306724                 LAMS > test > Pitanje 15 > Attempt   
306730                                        LAMS > test   
306734                                      LAMS > 217371   
306739                 LAMS > test > Pitanje 20 > Attempt   

        result.score.scaled  result.success  result.completion  
0                       NaN           False              False  
1                       NaN           False              False  
2                       NaN           False              False  
3                       NaN           False              False  
4                       NaN           False              False  
...                     ...             ...     

## Course

In [17]:
df[['Course', 'Institution']].drop_duplicates()

Course Institution
0                  Advanced Data Management Systems (THJ)         UEF
18463             Ihminen ja vuorovaikutteinen teknologia         UEF
153572  Проектиране на човеко-машинен интерфейс, летен...          SU
169434      Електронно обучение, летен семестър 2022/2023          SU
196531                       ARQUITECTURA DE COMPUTADORES          UL
271398                                   APLICACIONES WEB          UL
292147                                              CS120         BMU
300976                                              IT131         BMU

In [18]:
df['Course'].value_counts()

Course
Ihminen ja vuorovaikutteinen teknologia                              134855
ARQUITECTURA DE COMPUTADORES                                          74112
Електронно обучение, летен семестър 2022/2023                         26545
APLICACIONES WEB                                                      20699
Advanced Data Management Systems (THJ)                                18435
Проектиране на човеко-машинен интерфейс, летен семестър 2022/2023     15774
CS120                                                                  5263
IT131                                                                  3608
Name: count, dtype: int64

In [19]:
courses_dict = {
    'Ihminen ja vuorovaikutteinen teknologia': 'Human Factors of Interactive Technology',
    'Advanced Data Management Systems (THJ)': 'Advanced Data Management Systems',
    'Проектиране на човеко-машинен интерфейс, летен семестър 2022/2023': 'Human-computer interaction',
    'Електронно обучение, летен семестър 2022/2023': 'e-Learning',
    'ARQUITECTURA DE COMPUTADORES': 'Computer Architecture',
    'APLICACIONES WEB': 'Web Applications',
    'CS120': 'Computer Organization',
    'IT131': 'Computer Networks'
}

Note: BMU course name reference [here](https://www.metropolitan.ac.rs/files/2022/10/oas-informacione-tehnologije-2022.pdf)

In [20]:
df.replace({'Course': courses_dict}, inplace=True)

In [21]:
df

Institution                            Course actor.id  \
0              UEF  Advanced Data Management Systems    aeb0c   
1              UEF  Advanced Data Management Systems    aeb0c   
2              UEF  Advanced Data Management Systems    aeb0c   
3              UEF  Advanced Data Management Systems    aeb0c   
4              UEF  Advanced Data Management Systems    aeb0c   
...            ...                               ...      ...   
306723         BMU                 Computer Networks    87441   
306724         BMU                 Computer Networks    1c33c   
306730         BMU                 Computer Networks    98607   
306734         BMU                 Computer Networks    98607   
306739         BMU                 Computer Networks    00c5e   

                   timestamp    verb.id  \
0       2023-07-07T18:55:47Z     viewed   
1       2023-07-07T15:21:18Z     viewed   
2       2023-07-07T15:08:13Z     viewed   
3       2023-07-07T15:08:10Z     viewed   
4       2023-07-07T09:03:32Z     viewed   
...                      ...        ...   
306723  2023-06-11T22:57:27Z   answered   
306724  2023-05-16T00:05:45Z   answered   
306730  2023-05-12T22:10:50Z  completed   
306734  2023-02-27T22:45:33Z  completed   
306739  2023-06-01T21:23:16Z   answered   

                                   object.definition.name  \
0                  Advanced Data Management Systems (THJ)   
1                       Assignment Instructions: ER & EER   
2                                Recording: Intro lecture   
3                  Advanced Data Management Systems (THJ)   
4                  Advanced Data Management Systems (THJ)   
...                                                   ...   
306723                               IT131-T05-Pitanje 13   
306724                               IT131-T03-Pitanje 15   
306730                                          IT131-T03   
306734  Lekcija 2 - Osnove umrežavanja - Tradicionalna...   
306739                               IT131-T04-Pitanje 20   

                                   object.definition.type  \
0         http://id.tincanapi.com/activitytype/lms/course   
1                http://adlnet.gov/expapi/activities/link   
2                http://adlnet.gov/expapi/activities/link   
3         http://id.tincanapi.com/activitytype/lms/course   
4         http://id.tincanapi.com/activitytype/lms/course   
...                                                   ...   
306723  http://adlnet.gov/expapi/activities/cmi.intera...   
306724  http://adlnet.gov/expapi/activities/cmi.intera...   
306730     http://adlnet.gov/expapi/activities/assessment   
306734    http://id.tincanapi.com/activitytype/lms/module   
306739  http://adlnet.gov/expapi/activities/cmi.intera...   

                                                  context  \
0                                              UEF Moodle   
1       UEF Moodle > Advanced Data Management Systems ...   
2       UEF Moodle > Advanced Data Management Systems ...   
3                                              UEF Moodle   
4                                              UEF Moodle   
...                                                   ...   
306723                 LAMS > test > Pitanje 13 > Attempt   
306724                 LAMS > test > Pitanje 15 > Attempt   
306730                                        LAMS > test   
306734                                      LAMS > 217371   
306739                 LAMS > test > Pitanje 20 > Attempt   

        result.score.scaled  result.success  result.completion  
0                       NaN           False              False  
1                       NaN           False              False  
2                       NaN           False              False  
3                       NaN           False              False  
4                       NaN           False              False  
...                     ...             ...                ...  
306723                1.000            True            

## Add teaching type column

Specify "Project-based" / "Flipped classroom" teaching style

In [22]:
def get_teaching_type(course_name):
    if course_name in ['Computer Networks',
                       'Advanced Data Management Systems',
                        'e-Learning',
                        'Computer Architecture']:
        return 'Flipped classroom'
    else:
        return 'Project-based'

In [23]:
df['Teaching'] = df['Course'].apply(get_teaching_type)

In [24]:
df

Institution                            Course actor.id  \
0              UEF  Advanced Data Management Systems    aeb0c   
1              UEF  Advanced Data Management Systems    aeb0c   
2              UEF  Advanced Data Management Systems    aeb0c   
3              UEF  Advanced Data Management Systems    aeb0c   
4              UEF  Advanced Data Management Systems    aeb0c   
...            ...                               ...      ...   
306723         BMU                 Computer Networks    87441   
306724         BMU                 Computer Networks    1c33c   
306730         BMU                 Computer Networks    98607   
306734         BMU                 Computer Networks    98607   
306739         BMU                 Computer Networks    00c5e   

                   timestamp    verb.id  \
0       2023-07-07T18:55:47Z     viewed   
1       2023-07-07T15:21:18Z     viewed   
2       2023-07-07T15:08:13Z     viewed   
3       2023-07-07T15:08:10Z     viewed   
4       2023-07-07T09:03:32Z     viewed   
...                      ...        ...   
306723  2023-06-11T22:57:27Z   answered   
306724  2023-05-16T00:05:45Z   answered   
306730  2023-05-12T22:10:50Z  completed   
306734  2023-02-27T22:45:33Z  completed   
306739  2023-06-01T21:23:16Z   answered   

                                   object.definition.name  \
0                  Advanced Data Management Systems (THJ)   
1                       Assignment Instructions: ER & EER   
2                                Recording: Intro lecture   
3                  Advanced Data Management Systems (THJ)   
4                  Advanced Data Management Systems (THJ)   
...                                                   ...   
306723                               IT131-T05-Pitanje 13   
306724                               IT131-T03-Pitanje 15   
306730                                          IT131-T03   
306734  Lekcija 2 - Osnove umrežavanja - Tradicionalna...   
306739                               IT131-T04-Pitanje 20   

                                   object.definition.type  \
0         http://id.tincanapi.com/activitytype/lms/course   
1                http://adlnet.gov/expapi/activities/link   
2                http://adlnet.gov/expapi/activities/link   
3         http://id.tincanapi.com/activitytype/lms/course   
4         http://id.tincanapi.com/activitytype/lms/course   
...                                                   ...   
306723  http://adlnet.gov/expapi/activities/cmi.intera...   
306724  http://adlnet.gov/expapi/activities/cmi.intera...   
306730     http://adlnet.gov/expapi/activities/assessment   
306734    http://id.tincanapi.com/activitytype/lms/module   
306739  http://adlnet.gov/expapi/activities/cmi.intera...   

                                                  context  \
0                                              UEF Moodle   
1       UEF Moodle > Advanced Data Management Systems ...   
2       UEF Moodle > Advanced Data Management Systems ...   
3                                              UEF Moodle   
4                                              UEF Moodle   
...                                                   ...   
306723                 LAMS > test > Pitanje 13 > Attempt   
306724                 LAMS > test > Pitanje 15 > Attempt   
306730                                        LAMS > test   
306734                                      LAMS > 217371   
306739                 LAMS > test > Pitanje 20 > Attempt   

        result.score.scaled  result.success  result.completion  \
0                       NaN           False              False   
1                       NaN           False              False   
2                       NaN           False              False   
3                       NaN           False              False   
4                       NaN           False              False   
...                     ...             ...                ...   
306723                1.000            True     

In [25]:
df['Teaching'].value_counts()

Teaching
Project-based        176591
Flipped classroom    122700
Name: count, dtype: int64

## Object

In [26]:
df['object.definition.type'] = df['object.definition.type'].str.split('/').str[-1]

In [27]:
df

Institution                            Course actor.id  \
0              UEF  Advanced Data Management Systems    aeb0c   
1              UEF  Advanced Data Management Systems    aeb0c   
2              UEF  Advanced Data Management Systems    aeb0c   
3              UEF  Advanced Data Management Systems    aeb0c   
4              UEF  Advanced Data Management Systems    aeb0c   
...            ...                               ...      ...   
306723         BMU                 Computer Networks    87441   
306724         BMU                 Computer Networks    1c33c   
306730         BMU                 Computer Networks    98607   
306734         BMU                 Computer Networks    98607   
306739         BMU                 Computer Networks    00c5e   

                   timestamp    verb.id  \
0       2023-07-07T18:55:47Z     viewed   
1       2023-07-07T15:21:18Z     viewed   
2       2023-07-07T15:08:13Z     viewed   
3       2023-07-07T15:08:10Z     viewed   
4       2023-07-07T09:03:32Z     viewed   
...                      ...        ...   
306723  2023-06-11T22:57:27Z   answered   
306724  2023-05-16T00:05:45Z   answered   
306730  2023-05-12T22:10:50Z  completed   
306734  2023-02-27T22:45:33Z  completed   
306739  2023-06-01T21:23:16Z   answered   

                                   object.definition.name  \
0                  Advanced Data Management Systems (THJ)   
1                       Assignment Instructions: ER & EER   
2                                Recording: Intro lecture   
3                  Advanced Data Management Systems (THJ)   
4                  Advanced Data Management Systems (THJ)   
...                                                   ...   
306723                               IT131-T05-Pitanje 13   
306724                               IT131-T03-Pitanje 15   
306730                                          IT131-T03   
306734  Lekcija 2 - Osnove umrežavanja - Tradicionalna...   
306739                               IT131-T04-Pitanje 20   

       object.definition.type  \
0                      course   
1                        link   
2                        link   
3                      course   
4                      course   
...                       ...   
306723        cmi.interaction   
306724        cmi.interaction   
306730             assessment   
306734                 module   
306739        cmi.interaction   

                                                  context  \
0                                              UEF Moodle   
1       UEF Moodle > Advanced Data Management Systems ...   
2       UEF Moodle > Advanced Data Management Systems ...   
3                                              UEF Moodle   
4                                              UEF Moodle   
...                                                   ...   
306723                 LAMS > test > Pitanje 13 > Attempt   
306724                 LAMS > test > Pitanje 15 > Attempt   
306730                                        LAMS > test   
306734                                      LAMS > 217371   
306739                 LAMS > test > Pitanje 20 > Attempt   

        result.score.scaled  result.success  result.completion  \
0                       NaN           False              False   
1                       NaN           False              False   
2                       NaN           False              False   
3                       NaN           False              False   
4                       NaN           False              False   
...                     ...             ...                ...   
306723                1.000            True               True   
306724                0.000           False               True   
306730                0.375            True               True   
306734                  NaN           False              False   
306739                0.000           False               True   

                 Teaching  
0       Flipped classroom

In [28]:
df['object.definition.name'].describe()

count                                      299291
unique                                       1549
top       Ihminen ja vuorovaikutteinen teknologia
freq                                        34133
Name: object.definition.name, dtype: object

In [29]:
df['object.definition.name'].value_counts(normalize=True)[:20]

object.definition.name
Ihminen ja vuorovaikutteinen teknologia                              0.114046
ARQUITECTURA DE COMPUTADORES - Graduado en Ingeniería Informática    0.100688
Attempt                                                              0.049280
Електронно обучение, летен семестър 2022/2023                        0.036353
Проектиране на човеко-машинен интерфейс, летен семестър 2022/2023    0.035253
APLICACIONES WEB - Graduado en Ingeniería Informática                0.034381
Esimerkkejä käytettävyydestä ja käyttäjäkokemuksesta          0.029567
Ryhmätyöalue                                                       0.029339
Advanced Data Management Systems (THJ)                               0.022363
Tehtävä 3                                                          0.021013
Tehtävä 13                                                         0.013261
review                                                               0.007758
Exercise 2.1. Pipelining                 

A lot of the data is just the course name. Let's explore 'Attempt'.

In [76]:
df[df['object.definition.name'] == 'Attempt'][['object.definition.name', 'object.definition.type']].drop_duplicates()

object.definition.name object.definition.type
904                Attempt                attempt

In [30]:
df.drop(columns=['object.definition.name'], axis=1, inplace=True)

## Context

In [31]:
df['context'].value_counts(normalize=True)[:20]

context
UEF Moodle > Ihminen ja vuorovaikutteinen teknologia                                                                                                            0.153296
UEF Moodle                                                                                                                                                      0.136409
New Site                                                                                                                                                        0.135069
New Site > ARQUITECTURA DE COMPUTADORES - Graduado en Ingeniería Informática                                                                                    0.091302
ILEDA Project                                                                                                                                                   0.071606
UEF Moodle > Ihminen ja vuorovaikutteinen teknologia > Ryhmätyöalue                                                                              

The column mostly gives information about the institution / course.

In [32]:
df.drop(columns=['context'], axis=1, inplace=True)

## Result

In [33]:
df['result.score.scaled'].describe()

count    11558.00000
mean         0.78472
std          0.35866
min          0.00000
25%          0.70000
50%          1.00000
75%          1.00000
max          1.00000
Name: result.score.scaled, dtype: float64

In [34]:
df[~df['result.score.scaled'].isna()]['verb.id'].value_counts()

verb.id
completed    4309
scored       4235
answered     3014
Name: count, dtype: int64

In [35]:
df['result.success'].value_counts()

result.success
False    266458
True      32833
Name: count, dtype: int64

In [36]:
df[df['result.success'] == True]

Institution                            Course actor.id  \
74             UEF  Advanced Data Management Systems    6ddcf   
96             UEF  Advanced Data Management Systems    7f252   
97             UEF  Advanced Data Management Systems    710d9   
98             UEF  Advanced Data Management Systems    5df80   
99             UEF  Advanced Data Management Systems    94c89   
...            ...                               ...      ...   
306703         BMU                 Computer Networks    60c08   
306707         BMU                 Computer Networks    a0b51   
306715         BMU                 Computer Networks    dfea0   
306723         BMU                 Computer Networks    87441   
306730         BMU                 Computer Networks    98607   

                   timestamp    verb.id object.definition.type  \
74      2023-06-22T10:43:35Z     scored             assessment   
96      2023-06-21T20:54:34Z     scored             assessment   
97      2023-06-21T20:53:15Z     scored             assessment   
98      2023-06-21T20:52:48Z     scored             assessment   
99      2023-06-21T20:51:49Z     scored             assessment   
...                      ...        ...                    ...   
306703  2023-06-01T16:28:45Z   answered        cmi.interaction   
306707  2023-05-15T23:59:05Z   answered        cmi.interaction   
306715  2023-05-20T17:33:16Z   answered        cmi.interaction   
306723  2023-06-11T22:57:27Z   answered        cmi.interaction   
306730  2023-05-12T22:10:50Z  completed             assessment   

        result.score.scaled  result.success  result.completion  \
74                    0.600            True               True   
96                    0.950            True               True   
97                    1.000            True               True   
98                    0.950            True               True   
99                    1.000            True               True   
...                     ...             ...                ...   
306703                1.000            True               True   
306707                1.000            True               True   
306715                1.000            True               True   
306723                1.000            True               True   
306730                0.375            True               True   

                 Teaching  
74      Flipped classroom  
96      Flipped classroom  
97      Flipped classroom  
98      Flipped classroom  
99      Flipped classroom  
...                   ...  
306703  Flipped classroom  
306707  Flipped classroom  
306715  Flipped classroom  
306723  Flipped classroom  
306730  Flipped classroom  

[32833 rows x 10 columns]

In [37]:
df[df['result.success'] == True]['verb.id'].value_counts()

verb.id
answered     24305
completed     4301
scored        4227
Name: count, dtype: int64

In [38]:
df[df['result.success'] == True]['object.definition.type'].value_counts()

object.definition.type
cmi.interaction    24305
assessment          6735
quiz                1793
Name: count, dtype: int64

In [39]:
df['result.completion'].value_counts()

result.completion
False    257343
True      41948
Name: count, dtype: int64

In [40]:
df[df['result.completion'] == True]

Institution                            Course actor.id  \
74             UEF  Advanced Data Management Systems    6ddcf   
96             UEF  Advanced Data Management Systems    7f252   
97             UEF  Advanced Data Management Systems    710d9   
98             UEF  Advanced Data Management Systems    5df80   
99             UEF  Advanced Data Management Systems    94c89   
...            ...                               ...      ...   
306718         BMU                 Computer Networks    98607   
306723         BMU                 Computer Networks    87441   
306724         BMU                 Computer Networks    1c33c   
306730         BMU                 Computer Networks    98607   
306739         BMU                 Computer Networks    00c5e   

                   timestamp    verb.id object.definition.type  \
74      2023-06-22T10:43:35Z     scored             assessment   
96      2023-06-21T20:54:34Z     scored             assessment   
97      2023-06-21T20:53:15Z     scored             assessment   
98      2023-06-21T20:52:48Z     scored             assessment   
99      2023-06-21T20:51:49Z     scored             assessment   
...                      ...        ...                    ...   
306718  2023-05-12T22:10:50Z   answered        cmi.interaction   
306723  2023-06-11T22:57:27Z   answered        cmi.interaction   
306724  2023-05-16T00:05:45Z   answered        cmi.interaction   
306730  2023-05-12T22:10:50Z  completed             assessment   
306739  2023-06-01T21:23:16Z   answered        cmi.interaction   

        result.score.scaled  result.success  result.completion  \
74                    0.600            True               True   
96                    0.950            True               True   
97                    1.000            True               True   
98                    0.950            True               True   
99                    1.000            True               True   
...                     ...             ...                ...   
306718                0.000           False               True   
306723                1.000            True               True   
306724                0.000           False               True   
306730                0.375            True               True   
306739                0.000           False               True   

                 Teaching  
74      Flipped classroom  
96      Flipped classroom  
97      Flipped classroom  
98      Flipped classroom  
99      Flipped classroom  
...                   ...  
306718  Flipped classroom  
306723  Flipped classroom  
306724  Flipped classroom  
306730  Flipped classroom  
306739  Flipped classroom  

[41948 rows x 10 columns]

In [41]:
df[df['result.completion'] == True]['object.definition.type'].value_counts()

object.definition.type
cmi.interaction    33409
assessment          6744
quiz                1795
Name: count, dtype: int64

## Timestamp

In [42]:
df['timestamp']

0         2023-07-07T18:55:47Z
1         2023-07-07T15:21:18Z
2         2023-07-07T15:08:13Z
3         2023-07-07T15:08:10Z
4         2023-07-07T09:03:32Z
                  ...         
306723    2023-06-11T22:57:27Z
306724    2023-05-16T00:05:45Z
306730    2023-05-12T22:10:50Z
306734    2023-02-27T22:45:33Z
306739    2023-06-01T21:23:16Z
Name: timestamp, Length: 299291, dtype: object

In [45]:
def extract_from_date(timestamp):
        return (datetime.fromisoformat(timestamp[:-1]).year,
                datetime.fromisoformat(timestamp[:-1]).month,
                datetime.fromisoformat(timestamp[:-1]).day,
                datetime.fromisoformat(timestamp[:-1]).hour,
                datetime.fromisoformat(timestamp[:-1]).minute,
                datetime.fromisoformat(timestamp[:-1]).second)

In [47]:
def to_timestamp(tup):
    return pd.Timestamp(year=tup[0],
                        month=tup[1],
                        day=tup[2],
                        hour=tup[3],
                        minute=tup[4],
                        second=tup[5])

In [49]:
df['timestamp'] = df['timestamp'].apply(extract_from_date).apply(to_timestamp)

In [51]:
df

Institution                            Course actor.id  \
0              UEF  Advanced Data Management Systems    aeb0c   
1              UEF  Advanced Data Management Systems    aeb0c   
2              UEF  Advanced Data Management Systems    aeb0c   
3              UEF  Advanced Data Management Systems    aeb0c   
4              UEF  Advanced Data Management Systems    aeb0c   
...            ...                               ...      ...   
306723         BMU                 Computer Networks    87441   
306724         BMU                 Computer Networks    1c33c   
306730         BMU                 Computer Networks    98607   
306734         BMU                 Computer Networks    98607   
306739         BMU                 Computer Networks    00c5e   

                 timestamp    verb.id object.definition.type  \
0      2023-07-07 18:55:47     viewed                 course   
1      2023-07-07 15:21:18     viewed                   link   
2      2023-07-07 15:08:13     viewed                   link   
3      2023-07-07 15:08:10     viewed                 course   
4      2023-07-07 09:03:32     viewed                 course   
...                    ...        ...                    ...   
306723 2023-06-11 22:57:27   answered        cmi.interaction   
306724 2023-05-16 00:05:45   answered        cmi.interaction   
306730 2023-05-12 22:10:50  completed             assessment   
306734 2023-02-27 22:45:33  completed                 module   
306739 2023-06-01 21:23:16   answered        cmi.interaction   

        result.score.scaled  result.success  result.completion  \
0                       NaN           False              False   
1                       NaN           False              False   
2                       NaN           False              False   
3                       NaN           False              False   
4                       NaN           False              False   
...                     ...             ...                ...   
306723                1.000            True               True   
306724                0.000           False               True   
306730                0.375            True               True   
306734                  NaN           False              False   
306739                0.000           False               True   

                 Teaching  
0       Flipped classroom  
1       Flipped classroom  
2       Flipped classroom  
3       Flipped classroom  
4       Flipped classroom  
...                   ...  
306723  Flipped classroom  
306724  Flipped classroom  
306730  Flipped classroom  
306734  Flipped classroom  
306739  Flipped classroom  

[299291 rows x 10 columns]

## Actors

In [53]:
df['actor.id'].describe()

count     299291
unique       829
top        76d01
freq        6255
Name: actor.id, dtype: object

In [54]:
df['actor.id'].value_counts(normalize=True)

actor.id
76d01    0.020899
f231a    0.009623
d0853    0.007047
4722b    0.005830
d88b8    0.005700
           ...   
cd331    0.000003
47d0e    0.000003
7dd01    0.000003
297c3    0.000003
1ad31    0.000003
Name: proportion, Length: 829, dtype: float64

In [56]:
df['actor.id'] = pd.factorize(df['actor.id'])[0]

In [58]:
df

Institution                            Course  actor.id  \
0              UEF  Advanced Data Management Systems         0   
1              UEF  Advanced Data Management Systems         0   
2              UEF  Advanced Data Management Systems         0   
3              UEF  Advanced Data Management Systems         0   
4              UEF  Advanced Data Management Systems         0   
...            ...                               ...       ...   
306723         BMU                 Computer Networks       820   
306724         BMU                 Computer Networks       799   
306730         BMU                 Computer Networks       788   
306734         BMU                 Computer Networks       788   
306739         BMU                 Computer Networks       810   

                 timestamp    verb.id object.definition.type  \
0      2023-07-07 18:55:47     viewed                 course   
1      2023-07-07 15:21:18     viewed                   link   
2      2023-07-07 15:08:13     viewed                   link   
3      2023-07-07 15:08:10     viewed                 course   
4      2023-07-07 09:03:32     viewed                 course   
...                    ...        ...                    ...   
306723 2023-06-11 22:57:27   answered        cmi.interaction   
306724 2023-05-16 00:05:45   answered        cmi.interaction   
306730 2023-05-12 22:10:50  completed             assessment   
306734 2023-02-27 22:45:33  completed                 module   
306739 2023-06-01 21:23:16   answered        cmi.interaction   

        result.score.scaled  result.success  result.completion  \
0                       NaN           False              False   
1                       NaN           False              False   
2                       NaN           False              False   
3                       NaN           False              False   
4                       NaN           False              False   
...                     ...             ...                ...   
306723                1.000            True               True   
306724                0.000           False               True   
306730                0.375            True               True   
306734                  NaN           False              False   
306739                0.000           False               True   

                 Teaching  
0       Flipped classroom  
1       Flipped classroom  
2       Flipped classroom  
3       Flipped classroom  
4       Flipped classroom  
...                   ...  
306723  Flipped classroom  
306724  Flipped classroom  
306730  Flipped classroom  
306734  Flipped classroom  
306739  Flipped classroom  

[299291 rows x 10 columns]

In [59]:
df['actor.id'].value_counts(normalize=True)

actor.id
566    0.020899
87     0.009623
348    0.007047
605    0.005830
548    0.005700
         ...   
540    0.000003
541    0.000003
542    0.000003
543    0.000003
708    0.000003
Name: proportion, Length: 829, dtype: float64

In [60]:
df.to_csv('../data/processed.csv', index=False)